src/daft-warc/src/lib.rs

```diff
@@ -0,0 +1,575 @@
+pub enum WarcType {
+    Warcinfo,
+    Response,
+    Resource,
+    Request,
+    Metadata,
+    Revisit,
+    Conversion,
+    Continuation,
+    FutureType(String),
+}
+
+impl WarcType {
+    fn from_str(s: &str) -> Option<Self> {
+        match s.to_lowercase().as_str() {
+            "warcinfo" => Some(Self::Warcinfo),
+            "response" => Some(Self::Response),
+            "resource" => Some(Self::Resource),
+            "request" => Some(Self::Request),
+            "metadata" => Some(Self::Metadata),
+            "revisit" => Some(Self::Revisit),
+            "conversion" => Some(Self::Conversion),
+            "continuation" => Some(Self::Continuation),
+            _ => Some(Self::FutureType(s.to_string())),
+        }
+    }
+}
+
+impl std::fmt::Display for WarcType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Warcinfo => write!(f, "warcinfo"),
+            Self::Response => write!(f, "response"),
+            Self::Resource => write!(f, "resource"),
+            Self::Request => write!(f, "request"),
+            Self::Metadata => write!(f, "metadata"),
+            Self::Revisit => write!(f, "revisit"),
+            Self::Conversion => write!(f, "conversion"),
+            Self::Continuation => write!(f, "continuation"),
+            Self::FutureType(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+struct WarcHeaderState {
+    content_length: Option<usize>,
+    record_id: Option<Uuid>,
+    warc_date: Option<DateTime<Utc>>,
+    warc_type: Option<WarcType>,
+    header_lines: Vec<(String, String)>,
+}
+
+impl WarcHeaderState {
+    fn reset(&mut self) {
+        self.content_length = None;
+        self.record_id = None;
+        self.warc_date = None;
+        self.warc_type = None;
+        self.header_lines.clear();
+    }
+}
+
+struct WarcRecordBatchIterator {
+    reader: Box<dyn AsyncBufRead + Unpin + Send>,
+    schema: SchemaRef,
+    chunk_size: usize,
+    bytes_read: usize,
+    unprocessed_record: bool,
+    header_state: WarcHeaderState,
+    record_id_array: MutableUtf8Array<i64>,
+    warc_type_array: MutableUtf8Array<i64>,
+    warc_date_array: MutablePrimitiveArray<i64>,
+    warc_content_length_array: MutablePrimitiveArray<i64>,
+    content_array: MutableBinaryArray<i64>,
+    header_array: MutableUtf8Array<i64>,
+}
+
+impl WarcRecordBatchIterator {
+    fn new(
+        reader: Box<dyn AsyncBufRead + Unpin + Send>,
+        schema: SchemaRef,
+        chunk_size: usize,
+    ) -> Self {
+        Self {
+            reader,
+            schema,
+            chunk_size,
+            bytes_read: 0,
+            unprocessed_record: false,
+            header_state: WarcHeaderState {
+                content_length: None,
+                record_id: None,
+                warc_date: None,
+                warc_type: None,
+                header_lines: Vec::new(),
+            },
+            record_id_array: MutableUtf8Array::new(),
```

Comment:
We should also reserve capacities here that reflect the chunk size. Otherwise we will have many reallocations